### PR TITLE
Add new Blocks rich text editor (alpha)

### PR DESCRIPTION
**Pull request by simotae14**
This is the feature branch for work on the blocks rich text editor. Reviewed and QA'd PR's are merged into this branch.

### CODE REVIEW

**Comment by joshuaellis** [comment]
@simotae14 When is a good time to review this feature branch do you think?

**Comment by simotae14** [comment]
Hey @joshuaellis we need to prepare all the stuff for the alpha, so maybe it is better to add a review afterwards but the branches merged in the feature branch were already reviewed

**Comment by joshuaellis** [comment]
> Hey @joshuaellis we need to prepare all the stuff for the alpha, so maybe it is better to add a review afterwards but the branches merged in the feature branch were already reviewed

Can you let me know when is a good time to review the feature holistically please? :)

**Review by markkaylor — APPROVED** [review]
Good for me when all required tests pass

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+    }
+  };
+
+  return (
+    <>
+      <Select
+        startIcon={<Icon as={blocks[blockSelected].icon} />}
+        onChange={selectOption}
+        placeholder={blocks[blockSelected].label}
+        value={blockSelected}
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:271]: How come you need to control the state of this dropdown? Is it manipulated externally? 🤔

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+  const editor = useSlate();
+
+  const isActive = isBlockActive(editor, matchNode);
+  const isSelected = value === blockSelected;
+
+  React.useEffect(() => {
+    if (isActive && !isSelected) {
+      handleSelection(value);
+    }
+  }, [handleSelection, isActive, isSelected, value]);
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:305]: How come we have to do this? The SelectOption calls `onChange` when a new option is selected? This feels like unnecessarily boilerplate? Or if it's not maybe we should have a comment explaining this

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+const Wrapper = styled(Box)`
+  width: 100%;
+  max-height: 512px;
+  overflow: auto;
+  padding: ${({ theme }) => `${theme.spaces[3]} ${theme.spaces[4]}`};
+  font-size: ${({ theme }) => theme.fontSizes[2]};
+  background-color: ${({ theme }) => theme.colors.neutral0};
+  color: ${({ theme }) => theme.colors.neutral800};
+  line-height: ${({ theme }) => theme.lineHeights[6]};
+  border-radius: ${({ theme }) => theme.borderRadius};
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js:31]: Loads of these can be passed as props – how come we're not doing this?

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+      Editor.nodes(editor, {
+        at: Editor.unhangRange(editor, selection),
+        match: (node) => isListNode(node) && node.format === format,
+      })
+    );
+
+    return Boolean(match);
+  };
+
+  const isActive = isListActive();
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:356]: This feels weird, how come we're doing it this way?

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/BlocksInput/index.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      blocks.paragraph.handleEnterKey(editor);
+    }
+  };
+
+  const handleKeyDown = (event) => {
+    if (event.key === 'Enter') {
+      event.preventDefault();
+      handleEnter();
+    }
+  };
```
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/BlocksInput/index.js:70]: How come these are two separate functions?

**Review by joshuaellis — DISMISSED** [review]
I'm not sure you should update the design-system in a feature branch to be honest. It would be safer to do a separate PR into main and rebase – this means we don't need to revert your PR if there are issues etc.

Also i've left some notes, i think there's some weird over-engineering when using the select component particularly.

**Comment by markkaylor** [comment]
@joshuaellis thanks for the review but we will merge as is today for the stable/alpha release, we will have to handle your comments after the fact.

**Comment by joshuaellis** [comment]
Okay, you shouldn't update the DS in this PR – that's dangerous and if we have to revert for w/e reason we could loose your feature branch. This needs to be in it's own PR like every other team does.

**Comment by markkaylor** [comment]
> I'm not sure you should update the design-system in a feature branch to be honest

ok yes missed this part, we will handle it now

**Review by joshuaellis — APPROVED** [review]
might be worth squashing :)

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js (hunk 1 above)
Comment (simotae14) [reply packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:271]: Yes. I need to change the option selected everytime I select a portion of content inside the RTE editor, for example
if I select a portion of text that is an h3 I need to update immediately the option selected in the Dropdown to Heading 3
selection

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js (hunk 2 above)
Comment (simotae14) [reply packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:305]: this useEffect is to handle the case I mentioned before but you are right, I can provide a comment to explain why I added this part

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js (hunk 3 above)
Comment (simotae14) [reply packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js:31]: Ok, I am gonna create a new PR to handle your comments, thanks

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        <Flex direction="column" alignItems="stretch" gap={1}>
+          <Flex gap={1}>
+            <Typography variant="pi" fontWeight="bold" textColor="neutral800">
+              {label}
+              {required && <TypographyAsterisk textColor="danger600">*</TypographyAsterisk>}
+            </Typography>
+          </Flex>
+          <Slate
+            editor={editor}
+            initialValue={value || [{ type: 'paragraph', children: [{ type: 'text', text: '' }] }]}
```
Comment (madhurisandbhor) [inline packages/core/admin/admin/src/content-manager/components/BlocksEditor/index.js:81]: we can have this as a default value

File: packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js (hunk 4 above)
Comment (simotae14) [reply packages/core/admin/admin/src/content-manager/components/BlocksEditor/Toolbar/index.js:356]: We followed the Slate documentation in this case https://github.com/ianstormtaylor/slate/blob/main/site/examples/richtext.tsx you can find the isBlockIsActive which is similar
```
const isBlockActive = (editor, format, blockType = 'type') => {
  const { selection } = editor
  if (!selection) return false

  const [match] = Array.from(
    Editor.nodes(editor, {
      at: Editor.unhangRange(editor, selection),
      match: n =>
        !Editor.isEditor(n) &&
        SlateElement.isElement(n) &&
        n[blockType] === format,
    })
  )

  return !!match
}
```

**Pull request merged**